### PR TITLE
Add functions to start/end MSI transactions and check whether or not it is supported on the target machine

### PR DESCRIPTION
**Pull request by nirbar**

### CODE REVIEW

**Comment by rseanhall** [comment]
I'm not sure how useful these methods are for non-Burn applications. The two main benefits seem to be 1) making sure the transaction methods are actually available and 2) automatically setting the log to append.

Regarding 1, v4 doesn't compile native code to be compatible with XP anymore. I think we need to check what the minimum requirements are for the v142 toolset, it's possible that MSI v4.5 is always available.

For 2, I don't think configuring the log is an intrinsic requirement for ending an MSI transaction. It seems like something the caller should handle before calling this method. Even for Burn, it's possible to disable creating the log per package. In Burn, I think we should create a separate log file for the transaction itself.

I see in your branch name that you are referencing the original issue for supporting MSI transactions (https://github.com/wixtoolset/issues/issues/5386). Maybe we should reopen that issue if you don't think the feature was complete? In that case, you should probably update the WIP for the improvements you are planning.

**Comment by nirbar** [comment]
Windows Installer release notes say that it is available out of the box starting with Windows Vista SP2 and Windows Server 2008 SP2, so it is might not be available on Windows editions before that: https://support.microsoft.com/en-ca/help/942288/windows-installer-4-5-is-available

wiutil holds many functions that are only used by burn - WiuConfigureProductEx, WiuInstallProduct, WiuInitializeExternalUI and so forth

When not setting the log mode to append, the last log file in the transaction is overwritten which imposes a real problem in case of failures- you can't see the failure log file.
Even when logging is disabled for the transaction and log path is empty, not calling MsiEnableLog causes the last log to be overwritten. I guess you can call that a Windows Installer bug.

**Comment by rseanhall** [comment]
I'm having a hard time finding the minimum runtime requirements for the v142 toolset when creating a statically linked executable. If I understand things correctly, the v14x toolsets are built on top of the Universal CRT. The Universal CRT was first released in [KB2999226](https://support.microsoft.com/en-us/help/2999226/update-for-universal-c-runtime-in-windows), which had the minimum requirements of Vista SP2 and Server 2008 SP2. So with your findings it does appear we can rely on MSI v4.5.

That means that `WiuIsMsiTransactionSupported` would only be useful if someone built dutil with an older or deprecated toolset, targeting an OS that hasn't been supported by Microsoft for quite some time.

The wiutil functions you mentioned are merely wrappers around MSI functions, there's no Burn specific logic in them. Other MSI adjacent applications can use them without being constrained by Burn's philosophy. `WiuBeginTransaction` and `WiuEndTransaction` have Burn specific logic in them, and I don't agree with the logic you're including in them.

For the overwriting problem, it's specified clearly in the [`MsiEnableLog`](https://docs.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msienablelogw) documentation:

> The MsiEnableLog function sets the log mode for all subsequent installations that are initiated in the calling process.

It's Burn's fault for setting `MsiEnableLog` for the package and never undoing that.

**Comment by nirbar** [comment]
According to this link, only XP support was deprecated with v142: https://docs.microsoft.com/en-us/cpp/porting/features-deprecated-in-visual-studio

MsiEnableLog is indeed documented to apply for subsequent operations, yet in practice with MsiEndTransaction the commit / rollback actions are logged in the preceding log file while only few log lines are logged in the path specified in the MsiEnableLog call. 
If you don't call MsiEnableLog in append mode before MsiEndTransaction then the commit/rollback actions log overwrite the preceding log. If the previous install was not logged at all then calling MsiEnableLog before MsiEndTransaction has no effect on the previous log. Passing null/empty log path is possible in which case you'll get only the rollback/commit actions logged in the preceding log.
In that sense I think WiuEndTransaction works around MsiEnableLog / MsiEndTransaction behavior by preventing log overwrite

**Comment by rseanhall** [comment]
That link for XP support in v142 is misleading. XP was never supported in the v14x toolsets, you had to use the v14x_xp toolset. What that link is saying is that there is no v142_xp toolset. In order to target XP in VS2019, you have to use VS 2017's v141_xp toolset (which is available in the VS 2019 installer).

I think I finally understand why you want to call `MsiEnableLog` as part of those operations.

File: src/dutil/wiutil.cpp [hunk 1]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
     return hr;
 }
 
+extern "C" BOOL DAPI WiuIsMsiTransactionSupported(
+    )
+{
+    return (vpfnMsiBeginTransaction && vpfnMsiEndTransaction) ? TRUE : FALSE;
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:910]: This should just be `return vpfnMsiBeginTransaction && vpfnMsiEndTransaction`.

File: src/dutil/wiutil.cpp [hunk 2]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+    __in DWORD dwTransactionAttributes,
+    __out MSIHANDLE * phTransactionHandle,
+    __out HANDLE * phChangeOfOwnerEvent,
+    __in_z LPCWSTR szLogPath
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD er = ERROR_SUCCESS;
+
+    ExitOnNull((vpfnMsiBeginTransaction && vpfnMsiEndTransaction), hr, E_NOTIMPL, "Msi transactions are not supported");
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:924]: `ExitOnNull` should only be used to check if something is null. Other functions in this file are just doing `ExitFunction1(hr = E_NOTIMPL)` in this case. If you really want to log a message, then it should be

```
if (!WiuIsMsiTransactionSupported())
{
    ExitOnFailure(hr = E_NOTIMPL, "");
}
```

File: src/dutil/wiutil.cpp [hunk 3]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+
+    hr = WiuEnableLog(WIU_LOG_DEFAULT | INSTALLLOGMODE_VERBOSE | INSTALLLOGMODE_INSTALLSTART | INSTALLLOGMODE_INSTALLEND, szLogPath, INSTALLLOGATTRIBUTES_APPEND);
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:926]: There should be a parameter for the log mode so the caller decides what gets logged.

Also FYI, `INSTALLLOGMODE_INSTALLSTART` and `INSTALLLOGMODE_INSTALLEND` don't do anything for `MsiEnableLog`. They are for filtering messages in external UI handlers.

File: src/dutil/wiutil.cpp [hunk 4]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+    ExitOnFailure(hr, "Failed to enable logging for rollback boundary");
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:927]: This message should not reference rollback boundaries.

File: src/dutil/wiutil.cpp [hunk 5]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+
+extern "C" HRESULT DAPI WiuEndTransaction(
+    __in DWORD dwTransactionState,
+    __in_z LPCWSTR szLogPath
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD er = ERROR_SUCCESS;
+
+    ExitOnNull((vpfnMsiBeginTransaction && vpfnMsiEndTransaction), hr, E_NOTIMPL, "Msi transactions are not supported");
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:944]: See message above about `ExitOnNull`.

File: src/dutil/wiutil.cpp [hunk 6]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+
+    hr = WiuEnableLog(WIU_LOG_DEFAULT | INSTALLLOGMODE_VERBOSE | INSTALLLOGMODE_INSTALLSTART | INSTALLLOGMODE_INSTALLEND, szLogPath, INSTALLLOGATTRIBUTES_APPEND);
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:946]: See message above about new parameter.

File: src/dutil/wiutil.cpp [hunk 7]
```diff
@@ -886,6 +904,55 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
+    ExitOnFailure(hr, "Failed to enable logging for rollback boundary");
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:947]: See message about the contents of this message.

File: src/dutil/wiutil.cpp [hunk 8]
```diff
@@ -907,24 +907,33 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
 extern "C" BOOL DAPI WiuIsMsiTransactionSupported(
     )
 {
-    return (vpfnMsiBeginTransaction && vpfnMsiEndTransaction) ? TRUE : FALSE;
+    return (vpfnMsiBeginTransaction && vpfnMsiEndTransaction);
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:910]: Style nit: please don't use unnecessary parentheses.

We will be raising the minimum OS that we support to at least Vista SP2, so we can always rely on these functions being available. We don't need this method at all.

File: src/dutil/wiutil.cpp [hunk 9]
```diff
@@ -907,24 +907,33 @@ extern "C" HRESULT DAPI WiuSourceListAddSourceEx(
...
-    ExitOnNull((vpfnMsiBeginTransaction && vpfnMsiEndTransaction), hr, E_NOTIMPL, "Msi transactions are not supported");
+    if (!WiuIsMsiTransactionSupported())
+    {
+        ExitOnFailure(hr = E_NOTIMPL, "Msi transactions are not supported");
+    }
+
+    if (dwLogMode == 0)
+    {
+        dwLogMode = WIU_LOG_DEFAULT | INSTALLLOGMODE_VERBOSE;
+    }
```
Comment (rseanhall) [inline src/dutil/wiutil.cpp:933]: If `WiuEnableLog` doesn't have a default log mode then I don't think this function should either.

File: src/dutil/wiutil.cpp (hunk 8 above)
Comment (robmen) [reply src/dutil/wiutil.cpp:910]: Note: There are places in `dutil` that use function pointers to allow unittests to mock functions that change the OS. There may be value in continuing the follow that pattern.

**Comment by rseanhall** [comment]
Thanks!

**Comment by rseanhall** [comment]
I've already upgraded Burn to use this new version of dutil. If you need to do this in the future, you can use the updatepackage.ps1 in the Home repo like this:

```
.\updatepackage.ps1 -TargetFolder path\to\Burn -PackageName WixToolset.DUtil -NewVersion 4.0.59
```

**Pull request merged**